[#86759800] Netsim: connect to a router

islemaster opened the pull request:
In which I descend into callback madness.

Takes netsim from having -only- a simple lobby system to allowing the user to select a router, click 'connect' to establish a simulated connection between their node and the router (represented by a `NetSimWire` in a new table), have their node assigned a local address, and enter a view that shows a listing of hostnames/addresses on the router's local network.  That view will grow into our primary view for the simulator.

It turns out this transition is a real mess, and I could use some advice as I navigate through it.

Individual changes to look for:
- Routers are new.  A router **exists** if it has a row in the lobby table with 'router' type.  It is identified by its row ID.  Routers have a longer timeout than user rows.
- Wires are new.  A wire exists in a new table.  Its endpoints are nodes (lobby row IDs) and it contains metadata about those endpoints - hostnames, addresses, etc.
- `NetSimConnection` now has methods for connecting to a router, and can own a router and a wire.  The connection ticks both of these entities.
  - The wire is the simulated connection between the client node and the router.
  - The router reference really exists so that the connection can simulate the router's behavior as well.
  - Connecting to a router goes something like this:
    1. Optimistically create a wire between my node and the router.  Then check whether the router is over capacity; if it is, destroy my wire and reset.
    2. Ask the router to assign addresses on the new wire.  The client (acting as the router) scans the wires table to build an existing address table and find an open address.  The router then gives the open address to the wire, also setting its own address as 1.  (Routers are always 1 on their own network).
    3. Fire the connection status change event.
- New, rough UI for selecting a router and connecting to it, and a second 'widget' for displaying the router's local network.
- I separated the clean-up interval from the keep-alive interval.  Still hoping this will go away entirely, but it's a little less painful this way.

davidsbailey commented on apps/src/netsim/NetSimLobby.js:
```diff
@@ -79,28 +99,48 @@ NetSimLobby.createWithin = function (element, connection) {
   this.lobbyList_ = document.getElementById('netsim_lobby_list');
+
+  this.addRouterButton_ = document.getElementById('netsim_lobby_add_router');
+  dom.addClickTouchEvent(this.addRouterButton_,
```
Latest on jQuery: yes, we should move to using jQuery for things like this. caveat: mocha tests won't work with jQuery. My proposal: use jQuery here now, and when you go to add tests, there will be an extra step of making mocha pull in jQuery which Brent can help you accomplish.

davidsbailey commented on apps/src/netsim/netsimInstance.js:
```diff
@@ -0,0 +1,72 @@
+ * @param {!string} instanceID
+ * @returns {{getInstanceID: Function, getLobbyTable: Function, getWireTable: Function}}
+ */
+var netsimInstance = function (instanceID) {
```
Proposal: combine netsimInstance and netsimStorage into a single class, where everything in netsimStorage are static  members and everything in netsimInstance are instance members? I realize this might not actually make sense, but figured I'd propose it as a way to catalyze the process of me understanding the situation better.

islemaster replied:
Couter-proposal: netsimStorage should go away/be subsumed into Geoffrey's stuff in `shared/js/apps_api.js`and netsimInstance should just use that shared resource.

Confession: I still don't know how to use javascript in that shared/js folder.  I assume I just have to make it load everywhere to use it?  Seems not great.

davidsbailey replied:
I do not think the javascript that is in the shared/js folder and which accesses the /v3/apps API is used anywhere. It is just some test code that Geoffrey wrote and probably included (or copy/pasted) into an HTML file in order to execute it. 

Yes, please, replace it with a sharable API! We absolutely should have a client library for the shared-tables API. Nit regarding that: let's use CRUD terminology: create, read, readAll, update, delete.

davidsbailey replied:
If you are moving netsimStorage to /shared/js, then I propose renaming netsimInstance to netsimStorage.

davidsbailey commented on apps/src/netsim/NetSimNodeClient.js:
```diff
@@ -0,0 +1,257 @@
+ * @param {boolean} [autoReconnect=true]
+ */
+NetSimNodeClient.prototype.update = function (onComplete, autoReconnect) {
+  if (!onComplete) {
```
I would suggest getting rid of this pattern for setting onComplete, marking onComplete as a nullable parameter everywhere, and just calling it with "onComplete && onComplete()". Same for other callbacks which follow the same pattern.

islemaster replied:
Hmm... I made this change (using `onComplete && onComplete()`) but JSHint complains about it:
```
Expected an assignment or function call and instead saw an expression.
```
Should I change it back?  Adjust JSHint settings?

davidsbailey replied:
How about this? onComplete !== undefined && onComplete();

I would hesitate to go with something that jshint doesn't like so if you or
jshint don't like it then feel free to change it back... I was just looking
for ways to make the code a little denser.

On Thu, Feb 5, 2015 at 3:45 PM, Brad Buchanan <notifications@github.com>
wrote:

> In apps/src/netsim/NetSimNodeClient.js
> <https://github.com/code-dot-org/code-dot-org/pull/1207#discussion_r24211374>
> :
>
> > +  if (this.myWire) {
> > +    this.myWire.tick(clock);
> > +  }
> > +
> > +  if (this.myRouter) {
> > +    this.myRouter.tick(clock);
> > +  }
> > +};
> > +
> > +/**
> > + * If a client update fails, should attempt an automatic reconnect.
> > + * @param {function} [onComplete]
> > + * @param {boolean} [autoReconnect=true]
> > + */
> > +NetSimNodeClient.prototype.update = function (onComplete, autoReconnect) {
> > +  if (!onComplete) {
>
> Hmm... I made this change (using onComplete && onComplete()) but JSHint
> complains about it:
>
> Expected an assignment or function call and instead saw an expression.
>
> Should I change it back? Adjust JSHint settings?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/code-dot-org/code-dot-org/pull/1207/files#r24211374>.
>

islemaster replied:
Okay, I'm reverting that change.  I actually found the code got a little cleaner when I could assume the callback exists.  I'll do a pass to make sure I'm not providing a default handler where it makes no sense; e.g. in methods where the whole point is to asynchronously retrieve a resource.

davidsbailey commented on apps/src/netsim/periodicAction.js:
```diff
@@ -0,0 +1,109 @@
+      runLoop.tick.register(this, this.tick);
+    },
+
+    /**
```
I love the comments! For 1-liners, how about:
/** @param clock */

islemaster replied:
Done.

davidsbailey commented:
Hey Brad, great progress! I've added a few high-level comments. Please let me know when you're ready for another look.

davidsbailey commented:
Also, for your own sanity, after this PR is merged you might consider adding a test before adding new features. Just otherwise there's a chance other Studio Apps development will break netsim.

islemaster commented:
Hey @davidsbailey , I've moved/renamed
```
apps/src/netsim/netsimStorage.js => apps/src/appsApi.js, 
```
and renamed 
```
apps/src/netsim/netsimInstance.js => apps/src/netsim/NetSimTables.js
```

I also made sure appsApi includes the app storage and property bag interfaces (everything in @geoffrey-elliott 's tests).  Maybe this code is useful to you now as well?

I haven't put this in `shared/js` yet because we don't have a model yet for including that code in the grunt build process.  I experimented with this a little bit last night, but I feel like it's going to be a different step.  Hopefully making this code shareable between apps is a good first step to making it shareable across the whole application.

davidsbailey commented on apps/src/appsApi.js:
```diff
@@ -0,0 +1,213 @@
+  /** For the most part, property bags just use a different base URL. */
+  this.apiBaseUrl_ = '/v3/apps/' + app_publickey + '/shared-properties';
+};
+appsApi.SharedPropertyBag.prototype = Object.create(appsApi.AppStorage.prototype);
```
This inheritance relationship seems overly ambitious because a SharedPropertyBag should not have create, read, update and delete methods. I believe this is the "refused bequest" code smell: http://en.wikipedia.org/wiki/Code_smell

I would propose that the SharedPropertyBag "has an" AppStorage (rather than "is an").

islemaster replied:
Hmm, good call on this.  Should I do the same with SharedTable?  It *does* implement exactly the same methods, but is perhaps not a real child type.  Then again, it seems that AppStorage is more of a special case of SharedTable...

davidsbailey replied:
My other qualm here that a "post" is actually subtly different for a property bag than it is for a table, because with property bags it is interpreted as "create or update" whereas with the table the client is not allowed to "update" something which does not yet exist.

davidsbailey commented on apps/src/netsim/NetSimConnection.js:
```diff
@@ -67,28 +64,17 @@ var NetSimConnection = function (displayName, logger /*=new NetSimLogger(NONE)*/
-  /**
-   * Access object for instance lobby
-   * @type {netsimStorage.SharedStorageTable}
+   * Selected instance.
```
What is an "instance"?

islemaster replied:
:/ this keeps coming up, and I suppose I should change the terminology because of it.  I was hoping a reference to [this concept](http://en.wikipedia.org/wiki/Instance_dungeon) would be common enough to map.

davidsbailey commented:
LGTM after addressing above comments, which we elaborated on in HipChat.
